File: isegoria/src/main/java/com/eulersbridge/isegoria/feed/NewsDetailActivity.java
package com.eulersbridge.isegoria.feed;

import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.UiThread;
import android.support.v4.content.ContextCompat;
import android.support.v7.app.AppCompatActivity;
import android.text.TextUtils;
import android.view.View;
import android.widget.ImageView;
import android.widget.TextView;

import com.bumptech.glide.Priority;
import com.bumptech.glide.load.resource.drawable.DrawableTransitionOptions;
import com.eulersbridge.isegoria.GlideApp;
import com.eulersbridge.isegoria.Isegoria;
import com.eulersbridge.isegoria.R;
import com.eulersbridge.isegoria.common.BlurTransformation;
import com.eulersbridge.isegoria.common.Constant;
import com.eulersbridge.isegoria.common.TintTransformation;
import com.eulersbridge.isegoria.common.Utils;
import com.eulersbridge.isegoria.models.LikeInfo;
import com.eulersbridge.isegoria.models.NewsArticle;
import com.eulersbridge.isegoria.network.IgnoredCallback;
import com.eulersbridge.isegoria.network.SimpleCallback;

import org.parceler.Parcels;

import java.util.List;

import retrofit2.Response;

public class NewsDetailActivity extends AppCompatActivity {

    private Isegoria isegoria;

    private TextView likesTextView;
    private ImageView starView;

    private boolean userLikedArticle = false;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        setContentView(R.layout.news_detail_activity);

        getWindow().getDecorView().setSystemUiVisibility(View.SYSTEM_UI_FLAG_LAYOUT_STABLE
                    | View.SYSTEM_UI_FLAG_LAYOUT_FULLSCREEN);

        findViewById(R.id.back_button).setOnClickListener(view -> onBackPressed());

        ImageView articleImageView = findViewById(R.id.article_image);
        ImageView authorImageView = findViewById(R.id.article_image_author);
        likesTextView = findViewById(R.id.article_likes);
        starView = findViewById(R.id.article_star);

        Bundle extras = getIntent().getExtras();

        NewsArticle article = Parcels.unwrap(extras.getParcelable(Constant.ACTIVITY_EXTRA_NEWS_ARTICLE));
        populateTextContent(article);

        isegoria = (Isegoria)getApplication();

        isegoria.getAPI().getNewsArticleLikes(article.id).enqueue(new SimpleCallback<List<LikeInfo>>() {
            @Override
            protected void handleResponse(Response<List<LikeInfo>> response) {
                List<LikeInfo> likes = response.body();

                if (likes != null) populateLikes(likes);
            }
        });

        GlideApp.with(this)
                .load(article.getPhotoUrl())
                .priority(Priority.HIGH)
                .transforms(new BlurTransformation(NewsDetailActivity.this), new TintTransformation())
                .placeholder(R.color.lightGrey)
                .transition(DrawableTransitionOptions.withCrossFade())
                .into(articleImageView);

        String creatorPhotoURL = article.creator.profilePhotoURL;

        if (TextUtils.isEmpty(creatorPhotoURL)) {
            authorImageView.setVisibility(View.GONE);

        } else {
            GlideApp.with(this)
                    .load(creatorPhotoURL)
                    .priority(Priority.LOW)
                    .transition(DrawableTransitionOptions.withCrossFade())
                    .into(authorImageView);
        }
    }

    private void populateLikes(@NonNull List<LikeInfo> likes) {
        likesTextView.setText(String.valueOf(likes.size()));

        for (LikeInfo likeInfo : likes) {
            if (likeInfo.email.equals(isegoria.getLoggedInUser().email)) {
                userLikedArticle = true;
                starView.setImageResource(R.drawable.star);
                starView.setEnabled(true);
            }
        }
    }

    @SuppressWarnings("unchecked")
    @UiThread
    private void populateTextContent(final NewsArticle article) {

        TextView titleTextView = findViewById(R.id.article_title);
        titleTextView.setText(article.title);

        likesTextView.setText(String.valueOf(article.likeCount));

        TextView contentTextView = findViewById(R.id.article_content);
        contentTextView.setText(article.content);

        TextView authorNameTextView = findViewById(R.id.article_author);
        authorNameTextView.setText(article.creator.getFullName());

        TextView dateTextView = findViewById(R.id.article_date);
        dateTextView.setText(Utils.convertTimestampToString(this, article.dateTimestamp));

        final ImageView flagView = findViewById(R.id.article_flag);
        flagView.setOnClickListener(view -> flagView.setImageResource(R.drawable.flag));

        if (article.hasInappropriateContent) flagView.setImageResource(R.drawable.flagdefault);

        final ImageView starView = findViewById(R.id.article_star);
        starView.setOnClickListener(view -> {
            userLikedArticle = !userLikedArticle;

            IgnoredCallback callback = new IgnoredCallback<>();

            if (userLikedArticle) {
                isegoria.getAPI().likeArticle(article.id, isegoria.getLoggedInUser().email).enqueue(callback);

                starView.setColorFilter(ContextCompat.getColor(this, R.color.star_active));

                int newLikes = Integer.parseInt(String.valueOf(likesTextView.getText())) + 1;
                likesTextView.setText(String.valueOf(newLikes));

            } else {
                isegoria.getAPI().unlikeArticle(article.id, isegoria.getLoggedInUser().email).enqueue(callback);

                starView.setColorFilter(null);

                int newLikes = Integer.parseInt(String.valueOf(likesTextView.getText())) - 1;
                likesTextView.setText(String.valueOf(newLikes));
            }
        });

        //ImageView headShotImageView = (ImageView) rootView.findViewById(R.id.newsArticleHeadView);
        //network.getFirstPhotoImage(email, headShotImageView);

        //authorImageView.setVisibility(ViewGroup.VISIBLE);
    }
}

File: isegoria/src/main/kotlin/com/eulersbridge/isegoria/feed/news/detail/NewsDetailActivity.kt
package com.eulersbridge.isegoria.feed.news.detail

import android.arch.lifecycle.ViewModelProvider
import android.arch.lifecycle.ViewModelProviders
import android.os.Bundle
import android.view.View
import androidx.core.view.isGone
import com.bumptech.glide.Priority
import com.bumptech.glide.load.resource.drawable.DrawableTransitionOptions
import com.eulersbridge.isegoria.GlideApp
import com.eulersbridge.isegoria.R
import com.eulersbridge.isegoria.feed.news.ACTIVITY_EXTRA_NEWS_ARTICLE
import com.eulersbridge.isegoria.network.api.model.NewsArticle
import com.eulersbridge.isegoria.util.extension.ifTrue
import com.eulersbridge.isegoria.util.extension.observe
import com.eulersbridge.isegoria.util.extension.observeBoolean
import com.eulersbridge.isegoria.util.extension.toDateString
import com.eulersbridge.isegoria.util.transformation.BlurTransformation
import com.eulersbridge.isegoria.util.transformation.TintTransformation
import dagger.android.support.DaggerAppCompatActivity
import kotlinx.android.synthetic.main.news_detail_activity.*
import javax.inject.Inject

class NewsDetailActivity : DaggerAppCompatActivity() {

    @Inject
    lateinit var modelFactory: ViewModelProvider.Factory
    private lateinit var viewModel: NewsDetailViewModel

    private var userLikedArticle = false

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        setContentView(R.layout.news_detail_activity)

        window.decorView.systemUiVisibility = View.SYSTEM_UI_FLAG_LAYOUT_STABLE or View.SYSTEM_UI_FLAG_LAYOUT_FULLSCREEN

        backButton.setOnClickListener { onBackPressed() }

        val article = intent.getParcelableExtra<NewsArticle>(ACTIVITY_EXTRA_NEWS_ARTICLE)

        viewModel = ViewModelProviders.of(this, modelFactory)[NewsDetailViewModel::class.java]
        viewModel.setNewsArticle(article)
        createViewModelObservers()
    }

    private fun createViewModelObservers() {
        observe(viewModel.newsArticle) {
            it?.let {
                populateUIWithArticle(it)
            }
        }

        observe(viewModel.likeCount) {
            runOnUiThread { likesTextView.text = it.toString() }
        }

        ifTrue(viewModel.likedByUser) {
            runOnUiThread {
                userLikedArticle = true
                starImageView.setImageResource(R.drawable.star)
                starImageView.isEnabled = true
            }
        }
    }

    private fun populateUIWithArticle(article: NewsArticle) {
        GlideApp.with(this)
                .load(article.getPhotoUrl())
                .priority(Priority.HIGH)
                .transforms(BlurTransformation(this), TintTransformation())
                .placeholder(R.color.lightGrey)
                .transition(DrawableTransitionOptions.withCrossFade())
                .into(articleImageView)

        runOnUiThread {
            val creatorPhotoURL = article.creator.profilePhotoURL

            if (creatorPhotoURL.isNullOrBlank()) {
                authorImageView.isGone = true

            } else {
                GlideApp.with(this)
                        .load(creatorPhotoURL)
                        .priority(Priority.LOW)
                        .transition(DrawableTransitionOptions.withCrossFade())
                        .into(authorImageView)
            }

            titleTextView.text = article.title
            likesTextView.text = article.likeCount.toString()
            contentTextView.text = article.content
            authorNameTextView.text = article.creator.fullName
            dateTextView.text = article.date.toDateString(this)

            flagImageView.setOnClickListener { flagImageView.setImageResource(R.drawable.flag) }

            if (article.hasInappropriateContent)
                flagImageView.setImageResource(R.drawable.flag_default)

            starImageView.setOnClickListener { view ->
                userLikedArticle = !userLikedArticle

                view.isEnabled = false

                if (userLikedArticle) {
                    observeBoolean(viewModel.likeArticle()) { success ->
                        view.isEnabled = true

                        if (success) {
                            starImageView.setImageResource(R.drawable.star)

                            val newLikes = Integer.parseInt(likesTextView.text.toString()) + 1
                            likesTextView.text = newLikes.toString()
                        }
                    }

                } else {
                    observeBoolean(viewModel.unlikeArticle()) {success ->
                        view.isEnabled = true

                        if (success) {
                            starImageView.setImageResource(R.drawable.star_24dp)

                            val newLikes = Integer.parseInt(likesTextView.text.toString()) - 1
                            likesTextView.text = newLikes.toString()
                        }
                    }
                }
            }

            //ImageView headShotImageView = (ImageView) rootView.findViewById(R.id.newsArticleHeadView);
            //network.getFirstPhotoImage(email, headShotImageView);

            //authorImageView.setVisibility(ViewGroup.VISIBLE);
        }
    }
}
